File: frontend/src/services/companies.ts
import { apiGet } from "./api-client";

export type FinancialYear = {
  year: number; revenue: number | null; ebitda: number | null; earnings: number | null;
  ebitda_margin: number | null; operating_cash_flow: number | null; outstanding_shares: number | null;
};
export type CompanyPeer = { ticker: string; name: string; year: number; market_cap: number | null;
  pe_ttm: number | null; pb_mrq: number | null; revenue: number | null };
export type ApiCompany = {
  ticker: string; name: string; sector: string | null; country: string; tags: string[];
  price: number | null; price_date: string | null; daily_close_change: number | null;
  market_cap: number | null; employee_count: number | null; listing_date: string | null;
  website: string | null; esg_score: number | null;
  price_range_52w: { low: number | null; high: number | null };
  metrics: { eps: number | null; revenue_growth_yoy: number | null; earnings_growth_yoy: number | null;
    roe: number | null; revenue: number | null; ebitda: number | null; earnings: number | null;
    ebitda_margin: number | null; outstanding_shares: number | null; financial_year: number | null };
  financial_history: FinancialYear[];
  valuation: { forward_pe: number | null; intrinsic_value: number | null; pe: number | null;
    pb: number | null; enterprise_to_ebitda: number | null; year: number | null };
  mining: { name: string; operation: string | null; company_type: string | null; commodities: string[];
    activities?: string[]; site_count?: number | null; license_count?: number;
    licenses?: { activity: string | null; commodity: string | null; location: string | null; expiry_date: string | null }[] } | null;
  peers: CompanyPeer[];
  price_history: { date: string; close: number; volume: number | null }[];
  provenance: { company_report: string; mining_directory: string };
};

export const getCompany = (ticker: string) => apiGet<ApiCompany>(`/api/companies/${encodeURIComponent(ticker)}`);
export const getPeers = (ticker: string) => apiGet<CompanyPeer[]>(`/api/companies/${encodeURIComponent(ticker)}/peers`);
export const getContextGraph = (ticker: string) => apiGet<{ nodes: { id: string; type: string; label: string; detail?: string; source_reference?: string }[];
  edges: { source: string; target: string }[] }>(`/api/companies/${encodeURIComponent(ticker)}/context-graph`);

File: backend/app/api/companies.py
from fastapi import APIRouter, Depends
from app.dependencies import get_sectors
from app.schemas.common import Envelope, ResponseMeta
from app.services.company_service import CompanyService
from app.services.sectors_client import SectorsClient

router = APIRouter(tags=["companies"])


@router.get("/companies/{ticker}")
async def company(ticker: str, sectors: SectorsClient = Depends(get_sectors)):
    data, source, cached = await CompanyService(sectors).get_company(ticker)
    return Envelope(data=data, meta=ResponseMeta(source=source, cached=cached))


@router.get("/companies/{ticker}/peers")
async def peers(ticker: str, sectors: SectorsClient = Depends(get_sectors)):
    company_data, _, cached = await CompanyService(sectors).get_company(ticker)
    return Envelope(data=company_data["peers"], meta=ResponseMeta(source="sectors", cached=cached))


@router.get("/companies/{ticker}/context-graph")
async def context_graph(ticker: str, sectors: SectorsClient = Depends(get_sectors)):
    ticker = ticker.upper()
    company_data, source, cached = await CompanyService(sectors).get_company(ticker)
    nodes = [{"id": ticker.lower(), "type": "company", "label": company_data["name"], "detail": ticker}]
    mining = company_data.get("mining") or {}
    if mining:
        nodes.append({"id": "mining", "type": "mining", "label": mining["name"], "detail": mining.get("operation"),
                      "source_reference": company_data["provenance"]["mining_directory"]})
        for commodity in mining.get("commodities") or []:
            nodes.append({"id": f"commodity-{str(commodity).lower()}", "type": "commodity", "label": commodity,
                          "source_reference": company_data["provenance"]["mining_directory"]})
    if company_data["metrics"].get("revenue") is not None:
        nodes.append({"id": "financials", "type": "financial", "label": f"Financials {company_data['metrics']['financial_year']}",
                      "detail": "Revenue / EBITDA", "source_reference": company_data["provenance"]["company_report"]})
    for peer in company_data.get("peers", [])[:4]:
        nodes.append({"id": f"peer-{peer['ticker'].lower()}", "type": "peer", "label": peer["ticker"],
                      "detail": peer["name"], "source_reference": company_data["provenance"]["company_report"]})
    edges = [{"source": ticker.lower(), "target": n["id"]} for n in nodes[1:]]
    return Envelope(data={"nodes": nodes, "edges": edges}, meta=ResponseMeta(source=source, cached=cached,
        disclosure="Graph relationships are derived from Sectors company, mining, and peer fields."))
